astropy/utils/data_info.py

```diff
@@ -0,0 +1,303 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# Note: these functions and classes are tested extensively in astropy table
+# tests via their use in providing mixin column info, and in
+# astropy/tests/test_info for providing table and column info summary data.
+
+import os
+import sys
+from copy import deepcopy
+import weakref
+import numpy as np
+from functools import partial
+
+from ..extern import six
+from ..utils import OrderedDict
+from ..utils.compat import NUMPY_LT_1_8
+
+def data_info_factory(names, funcs):
+    """
+    Factory to create a function that can be used as an ``option``
+    for outputting data object summary information.
+
+    Examples
+    --------
+    >>> from astropy.utils.data_info import data_info_factory
+    >>> from astropy.table import Column
+    >>> c = Column([4., 3., 2., 1.])
+    >>> mystats = data_info_factory(names=['min', 'median', 'max'],
+    ...                             funcs=[np.min, np.median, np.max])
+    >>> c.info(option=mystats)
+    min = 1.0
+    median = 2.5
+    max = 4.0
+    n_bad = 0
+    length = 4
+
+    Parameters
+    ----------
+    names: list
+        List of information attribute names
+    funcs: list
+        List of functions that compute the corresponding information attribute
+
+    Returns
+    -------
+    func: function
+        Function that can be used as a data info option
+    """
+    def func(dat):
+        outs = []
+        for name, func in zip(names, funcs):
+            try:
+                if isinstance(func, six.string_types):
+                    out = getattr(dat, func)()
+                else:
+                    out = func(dat)
+            except:
+                outs.append('--')
+            else:
+                outs.append(str(out))
+
+        return OrderedDict(zip(names, outs))
+    return func
+
+
+def _get_data_attribute(dat, attr=None):
+    """
+    Get a data object attribute for the ``attributes`` info summary method
+    """
+    if attr == 'class':
+        val = type(dat).__name__
+    elif attr == 'dtype':
+        val = dat.info.dtype.name
+    elif attr == 'shape':
+        datshape = dat.shape[1:]
+        val = datshape if datshape else ''
+    else:
+        val = getattr(dat.info, attr)
+    if val is None:
+        val = ''
+    return str(val)
+
+
+class InfoDescriptor(object):
```

Comment:
> The one downside to the above approach is that the __init__ of the info class can't rely on _parent_ref.

I'm not sure what you mean exactly by "can't rely on", but definitely `_parent_ref` needs to be set every time `__get__` is called.  I had some issues along the way copying / transferring `info` and doing this on-the-fly parent referencing made things work better.  Apart from the bug with accessing `info` on the class object (I knew about this but forgot to address it), is there any problem in your approach of moving the `info._parent_ref = weakref.ref(instance)` statement outside the `if` block?

I also don't understand how the "ugly but perfectly valid" approach is different from the first.  I'm expecting another lesson in advanced Python.  :smile:

I guess it will be more elegant to be able to say `info = QuantifyInfo()` in the class definition, but we'll definitely need good code comments so mortals (and myself) of the future can understand what's happening.  I think my implementation is more easily understood by being very obvious about what is the info container and what is the descriptor.